Add a document for espresso-dev-node (#1722)
EspressoSystems/espresso-sequencer: doc/espresso-dev-node.md

# Espresso Dev Node

Espresso dev node is a node specifically designed for development and testing. It includes various nodes required to run
a complete Espresso network, such as `builder`, `sequencer`, etc. Developers can use it for development and testing.

## Download

We highly recommend you to use our Docker image. You can install it from the command line:

```cmd

> docker pull ghcr.io/espressosystems/espresso-sequencer/espresso-dev-node:main

Reviewer: Suggest putting the `docker run` command instead of the `pull` command.

```

## Parameters

> | Name                            | Type            | Environment Variable                 | Description                                                                                                                                                             |

Reviewer: We should also specify what the default behavior is if arguments are not provided.

| `dev_node_port`                 | `u16`           | `ESPRESSO_DEV_NODE_PORT`             | Port for the dev node. This is used to provide tools and information to facilitate developers debugging. If this is not provided, will use `20000` as the default port. |

## APIs

Once you have successfully run the dev node, you can access the corresponding ports to call the APIs of the
[`builder`](https://docs.espressosys.com/sequencer/api-reference/builder-api),
[`sequencer`](https://docs.espressosys.com/sequencer/api-reference/sequencer-api), and `prover`.

In addition, you can access the `dev_node_port` to retrieve debugging information. Here are the details of the dev node
API.

### GET /api/dev-info

This endpoint returns some debug information for you.

An example response is like this:

```json
{
  "builder_url": "http://localhost:41003/",
  "prover_port": 23156,
  "l1_url": "http://localhost:8545/",
  "light_client_address": "0xb075b82c7a23e0994df4793422a1f03dbcf9136f"
}
```

### POST /api/set-hotshot-down

> This endpoint sets the hotshot down at the given L1 height.

Reviewer: ```suggestion
This endpoint simulates the effect of a liveness failure of the hotshot consensus protocol in the Light Client smart contract.  
```

Let's try to be more precise.

By calling this, the L1 height in the light contract will be frozen, and rollups will detect the HotShot failure. This
is intended for testing rollups' functionalities when HotShot is down.

Parameters

| Name   | Type    | Description                              |
| ------ | ------- | ---------------------------------------- |
| height | integer | The L1 height from which hotshot is down |

### POST /api/set-hotshot-up

> This endpoint sets the hotshot up in the light client contract.

Reviewer: Similarly, let's be more precise and avoid using the terms "hotshut up/down" in the documentation.

# Espresso Dev Node

Espresso dev node is a node specifically designed for development and testing. It includes various nodes required to run
a complete Espresso network, such as `builder`, `sequencer`, etc. Developers can use it for development and testing.

## Download

We highly recommend you to use our Docker image. You can run it from the command line:

```cmd
docker run ghcr.io/espressosystems/espresso-sequencer/espresso-dev-node:main
```

## Parameters

| Name                            | Type            | Environment Variable                 | Default Value                                                 | Description                                                                                                                                  |
| ------------------------------- | --------------- | ------------------------------------ | ------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------- |

> | `rpc_url`                       | `Option<Url>`   | `ESPRESSO_SEQUENCER_L1_PROVIDER`     | Automatically launched Avil node if not provided.             | A JSON-RPC endpoint for the L1 to deploy to. If not provided, an Avil node will be launched automatically.                                   |

Reviewer 2: ```suggestion
| `rpc_url`                       | `Option<Url>`   | `ESPRESSO_SEQUENCER_L1_PROVIDER`     | Automatically launched Avil node if not provided.             | The JSON-RPC endpoint of the L1. If not provided, an Avil node will be launched automatically.                                   |
```

| `mnemonic`                      | `String`        | `ESPRESSO_SEQUENCER_ETH_MNEMONIC`    | `test test test test test test test test test test test junk` | Mnemonic for an L1 wallet. This wallet is used to deploy the contracts, so the account indicated by `ACCOUNT_INDEX` must be funded with ETH. |

> | `account_index`                 | `u32`           | `ESPRESSO_DEPLOYER_ACCOUNT_INDEX`    | `0`                                                           | Account index in the L1 wallet generated by `MNEMONIC` to use when deploying the contracts.                                                  |

Reviewer 2: ```suggestion
| `account_index`                 | `u32`           | `ESPRESSO_DEPLOYER_ACCOUNT_INDEX`    | `0`                                                           | Account index of the L1 wallet generated from `MNEMONIC`. Used when deploying contracts.                                                  |
```

| `builder_port`                  | `Option<u16>`   | `ESPRESSO_BUILDER_PORT`              | An unused port                                                | Port for connecting to the builder.                                                                                                          |
| `prover_port`                   | `Option<u16>`   | `ESPRESSO_PROVER_PORT`               | An unused port                                                | Port for connecting to the prover. If this is not provided, an available port will be selected.                                              |
| `dev_node_port`                 | `u16`           | `ESPRESSO_DEV_NODE_PORT`             | `20000`                                                       | Port for the dev node. This is used to provide tools and information to facilitate developers debugging.                                     |

## APIs

Once you have successfully run the dev node, you can access the corresponding ports to call the APIs of the
[`builder`](https://docs.espressosys.com/sequencer/api-reference/builder-api),
[`sequencer`](https://docs.espressosys.com/sequencer/api-reference/sequencer-api), and `prover`.

> In addition, you can access the `dev_node_port` to retrieve debugging information. Here are the details of the dev node

Reviewer 2: Consider linking to the toml specification or generated docs (if we generate documentation from the specification).

Author: The toml specification is used to generate an HTML doc for users accessing to a wrong route/endpoint. Of course we can also link the toml file here, but just a bit weird. And `builder api` page doesn't link anything here, so I just followed the conventions.

Reviewer 2: builder api documentation should also link to spec :smiley:  (IMO) .

API.

### GET /api/dev-info

This endpoint returns some debug information for you.

An example response is like this:

```json
{
  "builder_url": "http://localhost:41003/",
  "prover_port": 23156,
  "l1_url": "http://localhost:8545/",
  "light_client_address": "0xb075b82c7a23e0994df4793422a1f03dbcf9136f"
}
```

### POST /api/set-hotshot-down

This endpoint simulates the effect of a liveness failure of the hotshot consensus protocol in the Light Client smart
contract.

By calling this, the L1 height in the light contract will be frozen, and rollups will detect the HotShot failure. This
is intended for testing rollups' functionalities when HotShot is down.

> Parameters

Author: https://github.com/EspressoSystems/espresso-sequencer/pull/1722/commits/4e6f01c2f2684a6039db63ca0ea40730f32ba542

| Name   | Type    | Description                              |
| ------ | ------- | ---------------------------------------- |
| height | integer | The L1 height from which hotshot is down |

> ### POST /api/set-hotshot-up

Reviewer 2: I'm curious if we really need both endpoints. could we just have `post /api/hotshot` or similar which which reacted on the basis of current state of hotshot (like a toggle). Also since these are POSTs, it might be reasonable to move the command to the POST body. just a thought

Author: Yes that makes sense. I don't have any though about which is better. But the parameters of these 2 endpoints are totally different. I think we don't have to combine these into only 1.

Reviewer 2: Humm yea but it is POST, so if we do something like RPC we could just have a command input, and the request bodies would all differ based on the command. But anyway it not really import
